conda_forge_tick/migrators/staticlib.py

```diff
@@ -0,0 +1,581 @@
+
+BUILD_STRING_END_RE = re.compile(r".*_\d+$")
+
+logger = logging.getLogger(__name__)
+
+
+def _left_gt_right_rec(lrec, rrec):
+    lver = VersionOrder(lrec.version)
+    lbuild = lrec.build_number
+    rver = VersionOrder(rrec.version)
+    rbuild = rrec.build_number
+
+    if (lver > rver) or (lver == rver and lbuild > rbuild):
+        return True
+    else:
+        return False
+
+
+@lru_cache(maxsize=128)
+def _cached_match_spec(req: str) -> MatchSpec:
+    return MatchSpec(req)
+
+
+@lru_cache(maxsize=128)
+def _match_spec_is_exact(ms: MatchSpec) -> bool:
+    if (
+        ms.get_exact_value("name") is not None
+        and ms.get_exact_value("version") is not None
+        and ms.get_exact_value("build") is not None
+    ):
+        return True
+    else:
+        return False
+
+
+@lru_cache(maxsize=1)
+def _read_repodata(platform_arch: str) -> Any:
+    platform_arch = platform_arch.replace("_", "-")
+    rd_fn = fetch_repodata([platform_arch])[0]
+    with open(rd_fn) as fp:
+        rd = rapidjson.load(fp)
+    return rd
+
+
+@lru_cache(maxsize=128)
+def get_latest_static_lib(host_req: str, platform_arch: str) -> PackageRecord:
+    platform_arch = platform_arch.replace("_", "-")
+    rd = _read_repodata(platform_arch)
+    ms = _cached_match_spec(host_req)
+
+    max_rec = None
+    for key in ["packages", "packages.conda"]:
+        for fn, rec in rd[key].items():
+            if rec["name"] == ms.name:
+                rec = PackageRecord(**rec)
+                if ms.match(rec):
+                    if max_rec is None:
+                        max_rec = rec
+                    else:
+                        if _left_gt_right_rec(rec, max_rec):
+                            max_rec = rec
+
+    if max_rec is None:
+        raise ValueError(
+            f"Could not find a static lib for `{host_req}` on `{platform_arch}`!"
+        )
+    return max_rec
+
+
+@lru_cache(maxsize=128)
+def platform_has_rec(platform_arch: str, rec: PackageRecord) -> bool:
+    platform_arch = platform_arch.replace("_", "-")
+    rd = _read_repodata(platform_arch)
+    fn_dist = "-".join([rec.name, rec.version, rec.build])
+
+    for key in ["packages", "packages.conda"]:
+        for fn in rd[key]:
+            if fn_dist in fn:
+                return True
+    return False
+
+
+def extract_static_libs_from_meta_yaml_text(
+    meta_yaml_texts: tuple[str],
+    static_linking_host_requirement: str,
+    platform_arch: str | None = None,
+) -> set[tuple[str, PackageRecord]]:
+    ms = _cached_match_spec(static_linking_host_requirement)
+
+    logger.debug(
+        "computed ms '%s' for req '%s'",
+        ms.conda_build_form(),
+        static_linking_host_requirement,
+    )
+
+    recs = set()
+    for my in meta_yaml_texts:
+        for line in my.splitlines():
+            line = line.strip()
+            if line.startswith("-"):
+                line = line[1:].strip()
+                line = line.split("#", maxsplit=1)[0].strip()
```

Comment:
We've seen errors with this approach (e.g. lines like `- "https://domain.tld/path#anchor" # comment`). I guess this is ok for specs though, we shouldn't find those there.

Reply:
Yeah for sure. 🤞

Reply:
OMG yaml

```python
In [78]: yaml.load(" - mystr#", Loader=yaml.SafeLoader)
Out[78]: ['mystr#']

In [79]: yaml.load(" - mystr #", Loader=yaml.SafeLoader)
Out[79]: ['mystr']
```

Reply:
Also this:

```python
In [80]: yaml.load(" - 'mystr #'", Loader=yaml.SafeLoader)
Out[80]: ['mystr #']
```

Reply:
Here are some regex for this that cover list elements prepended by `-`:

```python
r"-\s?(?P<val>[^'\"\s]*)(?P<comment>\s+#[^'\"]*)?"
r"-\s?(?P<val>['][^']*['])(?P<comment>\s*#.*)?"
r"-\s?(?P<val>[\"][^\"]*[\"])(?P<comment>\s*#.*)?"
```

Reply:
fixed with new utility function that handles more edge cases, but not all edge cases.

Reply:
This is broken, so I punted on it.